static/app/components/searchQueryBuilder/tokens/filterKeyListBox/useFilterKeyListBox.tsx

```diff
@@ -1,64 +1,201 @@
 import clamp from 'sentry/utils/number/clamp';
 
 const MAX_OPTIONS_WITHOUT_SEARCH = 100;
 const MAX_OPTIONS_WITH_SEARCH = 8;
 
+function addRecentFiltersToItems({
+  items,
+  recentFilters,
+}: {
+  items: FilterKeyItem[];
+  recentFilters: string[];
+}): FilterKeyItem[] {
+  if (!recentFilters.length) {
+    return items;
+  }
+  return [
+    ...recentFilters.map(filter => ({
+      key: createRecentFilterOptionKey(filter),
+      value: filter,
+      textValue: filter,
+      type: 'recent-filter' as const,
+      label: filter,
+    })),
+    ...items,
+  ];
+}
+
+/**
+ * Finds the next item in the collection that matches the predicate.
+ * Iterates in the specified direction.
+ */
+function findNextMatchingItem(
+  state: ComboBoxState<FilterKeyItem>,
+  item: Node<FilterKeyItem> | null,
+  predicate: (item: Node<FilterKeyItem>) => boolean,
+  direction: 'after' | 'before'
+): Node<FilterKeyItem> | null {
+  let nextItem: Node<FilterKeyItem> | null = item;
+
+  do {
+    const nextKey = direction === 'after' ? nextItem?.nextKey : nextItem?.prevKey;
+    nextItem = nextKey ? state.collection.getItem(nextKey) : null;
+  } while (nextItem && !predicate(nextItem));
+
+  return nextItem;
+}
+
 export function useFilterKeyListBox({
   items,
   filterValue,
 }: {
   filterValue: string;
-  items: Array<KeyItem | KeySectionItem>;
+  items: FilterKeyItem[];
 }) {
   const {filterKeySections} = useSearchQueryBuilder();
   const [selectedSectionKey, setSelectedSection] = useState<Key | null>(null);
 
   const sections = useMemo(() => {
-    return items.filter(item => 'options' in item);
+    return items.filter(item => item.type === 'section');
   }, [items]);
 
-  const shownItems = useMemo(
-    () =>
-      items.filter(item => {
-        if (itemIsSection(item)) {
-          return !selectedSectionKey || item.key === selectedSectionKey;
+  const recentFilters = useRecentSearchFilters();
+
+  const filterKeyMenuItems = useMemo(() => {
+    const filteredByCategory = items.filter(item => {
+      if (itemIsSection(item)) {
+        return !selectedSectionKey || item.key === selectedSectionKey;
+      }
+
+      return true;
+    });
+
+    return addRecentFiltersToItems({items: filteredByCategory, recentFilters});
+  }, [items, recentFilters, selectedSectionKey]);
+
+  const customMenu: CustomComboboxMenu<FilterKeyItem> = props => {
+    return (
+      <FilterKeyListBox
+        {...props}
+        selectedSection={selectedSectionKey}
+        setSelectedSection={setSelectedSection}
+        sections={sections}
+        recentFilters={recentFilters}
+      />
+    );
+  };
+
+  const shouldShowExplorationMenu =
+    filterValue.length === 0 && filterKeySections.length > 0;
+
+  // This logic allows us to treat the recent filters as a separate section.
+  // If we have 5 recent filters, we want arrow up/down to cylce from the
+  // first recent filter to the first non-recent filter and vice versa.
+  const handleArrowUpDown = useCallback(
```

Comment:
Most of the changes are straightforward (adding items to the list and rendering them), but this was a bit hacky. By default, arrow up/down will take you through each of the recent filters. But since we render them horizontally, we  are modifying the arrow key behavior to make left/right go through the recent filters.

Reply:
Is there a limit to the quantity of recent filters? In theory this'd be fine if the user wants to use one of their recent filters for the query they're typing but if they don't they'd have to 'down'-press past them all to get to the list. I mean it's not that big a deal but more of a question (they can always type or use the cursor)

Reply:
Yeah, right now this only displays a max of 5 recent filters at the top (for space reasons). This arrow key behavior does help save a few keystrokes if they don't care about recent filters. And I also thought it seemed more natural behavior-wise since they are horizontal and in their own container